File: src/args.py
import argparse
import json
import os

def parse_args():
    parser = argparse.ArgumentParser(
        description="Gaze tracking with real-time or offline input and CSV export."
    )
    parser.add_argument(
        '--source', 
        type=str, 
        required=True, 
        choices=['webcam', 'image', 'video'],
        help="Data source for gaze detection: 'webcam', 'image', or 'video'."
    )
    parser.add_argument(
        '--path', 
        type=str, 
        help="Path to image or video if source != 'webcam'."
    )
    parser.add_argument(
        '--graph', 
        action='store_true', 
        help="Show a real-time graph of gaze coordinates."
    )
    parser.add_argument(
        '--affine', 
        action='store_true', 
        help="Apply affine transformation for face stabilization."
    )
    parser.add_argument(
        '--csv', 
        type=str, 
        help="Export gaze data to a CSV file."
    )
    parser.add_argument(
        '--csv_interval', 
        type=float, 
        default=1.0,
        help="Time interval in seconds for exporting CSV data."
    )
    return parser.parse_args()

def load_config(json_file_path):
    """
    Simplified config loader: no baseline references. We only handle:
      - source (webcam, image, video)
      - session_video, session_csv
      - export => { csv, graph, animation }
      - export_dir
      - dot_display, categorize, affine, etc.
    """
    if not os.path.exists(json_file_path):
        raise FileNotFoundError(f"Configuration file {json_file_path} does not exist.")
    
    with open(json_file_path, 'r') as file:
        config = json.load(file)

    # required 'source' check
    if 'source' not in config:
        raise ValueError("Config missing required 'source' field.")
    if config['source'] not in ['webcam', 'image', 'video']:
        raise ValueError("Invalid 'source'. Must be 'webcam', 'image', or 'video'.")

    # optional fields, default to None
    config.setdefault('session_video', None)
    config.setdefault('session_csv', None)

    # if user sets 'source' != 'webcam', we expect at least session_video or session_csv
    if config['source'] != 'webcam':
        if not (config['session_video'] or config['session_csv']):
            raise ValueError("Must provide 'session_video' or 'session_csv' when source != 'webcam'.")

    # Check mandatory 'dot_display' param
    if 'dot_display' not in config:
        raise ValueError("Config missing 'dot_display' boolean.")

    # handle 'export' dictionary
    if 'export' in config:
        export_opts = config['export']
        if not isinstance(export_opts, dict):
            raise ValueError("'export' must be a dictionary with 'csv' and 'graph' keys.")
        if 'csv' not in export_opts or 'graph' not in export_opts:
            raise ValueError("'export' dictionary must have 'csv' and 'graph' keys.")
        if (export_opts['csv'] or export_opts['graph'] or export_opts.get('animation',False)):
            if 'export_dir' not in config:
                raise ValueError("'export_dir' required if export.csv or export.graph is true.")

            if not os.path.isdir(config['export_dir']):
                os.makedirs(config['export_dir'], exist_ok=True)

    else:
        # no export info => default
        config['export'] = {'csv': False, 'graph': False, 'animation': False}

    # Additional booleans
    config.setdefault('affine', False)
    config.setdefault('categorize', False)

    return config

File: src/args.js
const fs = require('fs');
const path = require('path');

// Parse arguments from command line
function parseArgs() {
    const args = process.argv.slice(2);
    const options = {
        source: null,
        path: null,
        graph: false,
        affine: false,
        csv: null,
        csv_interval: 1.0
    };

    // Map arguments to options
    for (let i = 0; i < args.length; i++) {
        switch (args[i]) {
            case '--source':
                options.source = args[++i];
                if (!['webcam', 'image', 'video'].includes(options.source)) {
                    throw new Error("Invalid source type provided. Must be 'webcam', 'image', or 'video'.");
                }
                break;
            case '--path':
                options.path = args[++i];
                break;
            case '--graph':
                options.graph = true;
                break;
            case '--affine':
                options.affine = true;
                break;
            case '--csv':
                options.csv = args[++i];
                break;
            case '--csv_interval':
                options.csv_interval = parseFloat(args[++i]);
                break;
            default:
                throw new Error(`Unknown argument: ${args[i]}`);
        }
    }

    // Validation for required fields
    if (!options.source) {
        throw new Error("The '--source' argument is required.");
    }
    if (options.source !== 'webcam' && !options.path) {
        throw new Error("The '--path' argument is required when 'source' is not 'webcam'.");
    }

    return options;
}

// Load configuration from a JSON file
function loadConfig(jsonFilePath) {
    if (!fs.existsSync(jsonFilePath)) {
        throw new Error(`The configuration file ${jsonFilePath} does not exist.`);
    }

    const config = JSON.parse(fs.readFileSync(jsonFilePath, 'utf-8'));

    // Validate required fields
    const requiredFields = ['source'];
    requiredFields.forEach(field => {
        if (!(field in config)) {
            throw new Error(`The configuration is missing the required '${field}' field.`);
        }
    });

    if (!['webcam', 'image', 'video'].includes(config.source)) {
        throw new Error("Invalid value for 'source'. Must be 'webcam', 'image', or 'video'.");
    }

    // Set default values
    config.path = config.path || null;
    config.graph = config.graph ?? true;
    config.affine = config.affine ?? false;
    config.csv_interval = config.csv_interval ?? 1.0;
    config.categorize = config.categorize ?? false;

    if (config.source !== 'webcam' && !config.path) {
        throw new Error("The 'path' field is required when 'source' is not 'webcam'.");
    }

    if (!('dot_display' in config)) {
        throw new Error("The 'dot_display' parameter (boolean) is not provided.");
    }

    // Handle 'export' sub-requirements
    if (config.export) {
        const exportOptions = config.export;
        if (typeof exportOptions !== 'object' || !('csv' in exportOptions) || !('graph' in exportOptions)) {
            throw new Error("The 'export' field must be an object containing 'csv' and 'graph' keys.");
        }

        if (exportOptions.csv || exportOptions.graph || exportOptions.animation) {
            if (!config.export_dir) {
                throw new Error("The 'export_dir' field is required when 'export.csv' or 'export.graph' is enabled.");
            }

            const exportDir = config.export_dir;
            if (!fs.existsSync(exportDir)) {
                fs.mkdirSync(exportDir, { recursive: true });
            }

            if (exportOptions.csv) {
                config.csv = path.join(exportDir, "raw_data.csv");
                console.log(`CSV export will be saved to: ${config.csv}`);
            } else {
                config.csv = null;
            }

            if (exportOptions.graph) {
                config.graph_out = path.join(exportDir, "final_comprehensive_plots.png");
                console.log(`Graph export will be saved to: ${config.graph_out}`);
            } else {
                config.graph_out = null;
            }

            if (exportOptions.animation) {
                config.animation_out = path.join(exportDir, "animation.mp4");
            }
        } else {
            config.csv = null;
            config.graph_out = null;
        }
    } else {
        config.csv = null;
        config.graph_out = null;
    }

    return config;
}

// Example usage
try {
    const options = parseArgs();
    console.log("Parsed command-line arguments:", options);

    const config = loadConfig('./config.json');  // Replace with actual path to JSON file
    console.log("Loaded configuration:", config);
} catch (error) {
    console.error(error.message);
}
